src/protocols/secure_channel/StatusReport.h

```diff
@@ -0,0 +1,101 @@
+/**
+ *    @file
+ *      Defines the StatusReport class and General Codes for StatusReport messages.
+ */
+
+#pragma once
+
+#include <support/BufferWriter.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace Protocols {
+namespace SecureChannel {
+
+/**
+ *  Encapsulates the data included in a StatusReport message, and provides methods for writing to and reading from PacketBuffers
+ *  that contain StatusReport messages.
+ */
+class DLL_EXPORT StatusReport
+{
+public:
+    /**
+     *  Construct a StatusReport with zero-d out fields (for use before calling \c Parse() ).
+     */
+    StatusReport();
+
+    /**
+     *  Construct a StatusReport with no additional ProtocolData.
+     *
+     *  @param generalCode Required, one of the \c GeneralStatusCode values listed in \c secure_channel/Constants.h
+     *  @param protocolId Must specify a ProtocolId which consists of Vendor Id (upper 16 bits) and ProtocolId (lower 16 bits)
+     *  @param protocolCode A code defined by the specified protocol which provides more information about the status
+     */
+    StatusReport(uint16_t generalCode, uint32_t protocolId, uint16_t protocolCode);
+
+    //
+    /**
+     *  Construct a StatusReport with additional ProtocolData.
+     *
+     *  WARNING: The caller MUST ensure that the data referenced by protocolData not be freed before calling \c WriteToBuffer()
+     *      This function will not make a copy of the data.
```

Comment:
Wouldn't it be better to take a PacketBufferHandle rather than protocolData pointer + datalen?

Reply:
Why would this be better? Do you think the protocol-specific data will often be stored in a PacketBuffer already? If not, is it a good idea to make the caller allocate a PacketBuffer just to construct a StatusReport (which will already require a second PacketBuffer later for writing the message).

Reply:
we discussed offline and I have since made this change to use a PacketBufferHandle instead of pointer+size